### DISCUSSION TITLE
Combined ViewSet

### DISCUSSION THREAD

**Discussion by pomali** [encode/django-rest-framework#9398] [General] [answered]
Hi,
for a few days I was trying to come up with good solution to having multiple ViewSets with different serializers, permissions, etc. on the same url prefix (using router).

One solution is to override `get_serializers`, `get_permissions`, `filter_queryset` based on `self.action`, but solution that I imagined that would work was writing multiple ViewSets and combining them into one router url prefix.

I tried to create metaclass that would do this combining, like this

```python
# views.py
class MyCombinedViewSet(
  MyOwnerActionsViewSet,
  MyReadViewSet, .....
  metaclass=CombinedViewSetMetaclass
):
    pass
    
 # urls.py
router.register(r"my-api", MyCombinedViewSet, "my-api")
```

I have somewhat working version, but now I hit a bump with correctly creating as_view()/dispatch() and started thinking that in the end it may be bad idea. 

**What are your thoughts?** How bad idea is it?

More usecases:
- having permissions/serializers/... organized in classes
- having retrieve my (current user's) object instead of list (GET /my-api/) while creating/updating works there too (POST/PUT /my-api/)

**Comment by auvipy** [comment, marked as answer]
you can do it in your project or in a separate package. DRF do not intend to start new patterns at the stage of it lifecycle.

**Reply by pomali** [reply]
I agree with that, but I was curious about your thoughts. It seemed like the natural progression from ViewSets, but as tomchristie  mentioned ViewSests can be a bit unclear, and this would make it even worse.

**Reply by auvipy** [reply]
viewsets has their own positive sides and some downsides too so it depends on solely your needs

**Comment by lovelydinosaur** [comment]
Feedback... I'm not actually all that keen on ViewSet full stop. I think they make for really unclear code, and were arguably a mistake.

So, for my part, no I don't think it's likely to be a fantastic approach.

**Reply by pomali** [reply]
that is interesting, so preferred solution is using generic views and manually binding them to urls?

**Reply by gegoune** [reply]
This is very interesting point which many might be unaware of. Do you think documentation should discourage users from using `ViewSet`s?

**Reply by lovelydinosaur** [reply]
Eh, lots of other people would prob disagree with me, and find using them helpful so... 🤷🏼‍♂️
I don't particularly see us changing the documentation radically.

**Reply by auvipy** [reply]
Tom, viewsets has its own merits / positive sides as well. it depends on the need of the project / endpoint. let us not fully discourage it. those who need more higher level pattern, can build their own, out of the DRF package's scope. My 2 cents
